crates/extensions/c8y_mapper_ext/src/availability.rs

```diff
@@ -0,0 +1,255 @@
+    heartbeat: &HeartbeatPayload,
+) -> Option<MqttMessage> {
+    let health_topic_name = heartbeat.health.clone();
+
+    match map.get(&health_topic_name) {
+        Some(health_status) if health_status.status == "up" => {
+            Some(MqttMessage::new(c8y_topic, "{}"))
+        }
+        _ => {
+            debug!("Heartbeat message is not generated because the status of '{health_topic_name}' is not 'up'");
+            None
+        }
+    }
+}
+
+pub async fn set_heartbeat_timer_from_metadata(
+    period: i16,
+    status_map: &mut HashMap<TopicWithoutPrefix, HealthStatus>,
+    timer_map: &mut HashMap<EntityTopicId, TopicWithoutPrefix>,
+    metadata: &EntityMetadata,
+    timer_sender: LoggingSender<SyncStart>,
+) -> Result<(), ChannelError> {
+    if period < 0 {
+        return Ok(());
+    }
+
+    let interval: u64 = period.try_into().unwrap(); // Already checked if it's positive number
+    let topic_id = metadata.topic_id.clone();
+    match get_lead_service_topic(metadata) {
+        Ok(health_topic) => {
+            set_heartbeat_timer(
+                interval,
+                status_map,
+                timer_map,
+                &topic_id,
+                health_topic,
+                timer_sender,
+            )
+            .await?;
+        }
+        Err(err) => {
+            info!("Failed to set a timer for the device '{topic_id}' to send its availability heartbeat due to: {err}");
+        }
+    }
+
+    Ok(())
+}
+
+pub async fn set_heartbeat_timer(
+    interval: u64,
+    status_map: &mut HashMap<TopicWithoutPrefix, HealthStatus>,
+    timer_map: &mut HashMap<EntityTopicId, TopicWithoutPrefix>,
+    topic_id: &EntityTopicId,
+    health_topic: TopicWithoutPrefix,
+    timer_sender: LoggingSender<SyncStart>,
+) -> Result<(), ChannelError> {
+    insert_new_health_status_entry(status_map, &health_topic);
+    start_heartbeat_timer(
+        timer_sender.clone(),
+        interval,
+        topic_id.clone(),
+        health_topic.clone(),
+    )
+    .await?;
+    update_active_heartbeat_timer_map(timer_map, topic_id, &health_topic);
+
+    Ok(())
+}
+
+pub async fn start_heartbeat_timer(
+    mut timer_sender: LoggingSender<SyncStart>,
+    interval: u64,
+    device_entity: EntityTopicId,
+    health_topic: TopicWithoutPrefix,
+) -> Result<(), ChannelError> {
+    let heartbeat_payload = HeartbeatPayload {
+        device: device_entity,
+        health: health_topic,
+    };
+
+    timer_sender
+        .send(SyncStart::new(
+            Duration::from_secs(interval * 60), // interval is in minutes
+            TimeoutKind::Heartbeat(heartbeat_payload),
+        ))
+        .await?;
+
+    Ok(())
+}
+
+pub fn get_lead_service_topic(entity: &EntityMetadata) -> Result<TopicWithoutPrefix, String> {
+    match entity.other.get("@health") {
+        Some(maybe_topic_name) => match maybe_topic_name.as_str() {
+            Some(topic_name) if Topic::new(topic_name).is_ok() => Ok(topic_name.to_string()),
```

Comment:
I feel we should validate that the value is a valid entity topic id. I'd have validated that against the entity store as well, but may be that's too early, as the lead service may not be registered by then.

Reply:
The lead service can be anything, that's why the type is not `EntityTopicId` and also no querying to entity store.